server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java

```diff
@@ -379,6 +430,45 @@ public KeywordFieldMapper build(MapperBuilderContext context) {
                 this
             );
         }
+
+        private FieldType fieldType(final IndexSortConfig indexSortConfig, final IndexMode indexMode, final String fullFieldName) {
+            if (shouldUseDocValuesSparseIndex(indexSortConfig, indexMode, fullFieldName)) {
+                return new FieldType(Defaults.FIELD_TYPE_WITH_SKIP_DOC_VALUES);
+            }
+            return new FieldType(Defaults.FIELD_TYPE);
+        }
+
+        private boolean shouldUseDocValuesSparseIndex(
+            final IndexSortConfig indexSortConfig,
+            final IndexMode indexMode,
+            final String fullFieldName
+        ) {
+            return (defaultIndexedAndDocValues() || isNotIndexedAndHasDocValues())
```

Comment:
Here I also enable the sparse index if the user explicitly defines the field to be indexed and with doc values. In that case we pay a little price in terms of storage footprint to take advantage of the sparse index.